### PR TITLE
Update Go tests

### LINKED ISSUE #231: Use `errors.Is()` in Golang tests
Instead of using the `assert.Error()` function, which checks that an error occurred, in a future where our validation functions return more errors, we might also use `errors.Is()` to check that we get the same error back.

_Originally posted by @pxp928 in https://github.com/in-toto/attestation/pull/220#discussion_r1205883501_

### PR DESCRIPTION
**Pull request by adityasaky**
Fixes #231 

* Create named errors for reuse and testing
* Assert in tests expected named errors
* Add test for subject-specific restriction on subject RD

### CODE REVIEW

File: go/v1/statement_test.go [hunk 1]
```diff
@@ -84,7 +84,20 @@ func TestBadStatementSubject(t *testing.T) {
 
 	err = got.Validate()
 
-	assert.Error(t, err, "Error: created malformed Statement (bad subject)")
+	// With this input, we encounter the RD object's validation error
+	assert.ErrorIs(t, err, ErrRDRequiredField, "Error: created malformed Statement (bad subject)")
```
Comment (TomHennen) [inline go/v1/statement_test.go:88]: optional: I'm wondering if these should be separate tests?  Alternatively maybe this should be a table driven test? https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

Comment (TomHennen) [inline go/v1/statement_test.go:88]: nit: can the error message between these tests distinguish which field is the problem?  (This and the following test have the same messages).  Maybe resolved if they're separate tests?

**Review by TomHennen — CHANGES_REQUESTED** [review]
Looks good!  Just a couple comments.

File: go/v1/statement_test.go (hunk 1 above)
Comment (adityasaky) [reply go/v1/statement_test.go]: Updated to use table, I was trying to keep the diff small but this is def better.

Comment (adityasaky) [reply go/v1/statement_test.go]: Done!

**Review by TomHennen — APPROVED** [review]
Thanks!

**Review by marcelamelara — APPROVED** [review]
Thank you @adityasaky !

**Pull request merged**